### PR TITLE
[Merged by Bors] - feat: add instance `SubringClass.nonUnitalSubringClass`

### PR DESCRIPTION
**Pull request by j-loreaux**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
!bench

**Comment by j-loreaux** [comment]
!bench

**Comment by j-loreaux** [comment]
!bench

File: Mathlib/RingTheory/NonUnitalSubring/Basic.lean [hunk 1]
```diff
@@ -77,8 +77,8 @@ section NonUnitalSubringClass
 
 /-- `NonUnitalSubringClass S R` states that `S` is a type of subsets `s ⊆ R` that
 are both a multiplicative submonoid and an additive subgroup. -/
-class NonUnitalSubringClass (S : Type*) (R : Type u) [NonUnitalNonAssocRing R]
-    [SetLike S R] extends NonUnitalSubsemiringClass S R, NegMemClass S R : Prop where
+class abbrev NonUnitalSubringClass (S : Type*) (R : Type u) [NonUnitalNonAssocRing R]
+    [SetLike S R] : Prop := NonUnitalSubsemiringClass S R, NegMemClass S R
```
Comment (eric-wieser) [inline Mathlib/RingTheory/NonUnitalSubring/Basic.lean:81]: This looks like a bad idea to me; `class abbrev` creates loops where it both implies and is implied by the nested classes.

**Comment by eric-wieser** [comment]
Your original commit looks like the right one to me (though the instance should be called `toNonUnitalSubringClass`)

File: Mathlib/RingTheory/NonUnitalSubring/Basic.lean (hunk 1 above)
Comment (j-loreaux) [reply Mathlib/RingTheory/NonUnitalSubring/Basic.lean:81]: I know, I was just experimenting given the terrible performance in the earlier commits.

**Comment by xroblot** [comment]
For the file `Discriminant.lean`, it seems that adding the line
```
 have : CharZero K := SubsemiringClass.instCharZero K
```
before lines 340, 383 and 422 helps with the following line
```
haveI : NumberField K := @NumberField.mk _ _ inferInstance hK₀
```
Also, the `simp` on lines 363 and 407 are slow and in both cases the line
```
      convert hx₁ <;> simp
```
can be replaced with 
```
      convert hx₁
      · simp only [IntermediateField.lift_top]
      · simp only [IntermediateField.lift_adjoin, Set.image_singleton]
```
for a significative improvement.

**Comment by j-loreaux** [comment]
!bench

**Comment by j-loreaux** [comment]
!bench

**Comment by j-loreaux** [comment]
!bench

**Comment by j-loreaux** [comment]
!bench

**Comment by j-loreaux** [comment]
!bench

**Pull request closed without merging**